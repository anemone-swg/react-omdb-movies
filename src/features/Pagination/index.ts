export { default as Pagination } from "./ui/Pagination";
export { paginationReducer, paginationActions } from "./model/slice";
export { selectPage } from "./model/selectors";
