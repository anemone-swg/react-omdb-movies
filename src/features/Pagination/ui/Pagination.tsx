import type { JSX } from "react";
import { selectPage } from "../model/selectors";
import { paginationActions } from "../model/slice";
import { useAppSelector } from "@/shared/lib/hooks/useAppSelector";
import { useAppDispatch } from "@/shared/lib/hooks/useAppDispatch";
import { useTranslation } from "react-i18next";

/**
 * Props компонента Pagination.
 *
 * @property {string | undefined} totalResults - Данные с поиска фильмов.
 * @property {boolean} [isFetching] - Флаг загрузки данных.
 */

export interface PaginationProps {
  totalResults: string | undefined;
  isFetching?: boolean;
}

/**
 * React-компонент для пагинации.
 *
 * @component
 * @param {PaginationProps} props - Props компонента.
 * @returns {JSX.Element} JSX-элемент с найденными фильмами.
 */
const Pagination = ({
  totalResults,
  isFetching = false,
}: PaginationProps): JSX.Element => {
  const { t } = useTranslation();
  const dispatch = useAppDispatch();
  const page = useAppSelector(selectPage);

  return (
    <>
      {!isFetching && totalResults && Number(totalResults) > 0 && (
        <div className="flex justify-center gap-4 mt-6">
          <button
            className="bg-button text-white px-4 py-2 rounded transition-colors disabled:opacity-50"
            disabled={page === 1 || isFetching}
            onClick={() => dispatch(paginationActions.setPage(page - 1))}
          >
            {t("назад")}
          </button>
          <span className="px-4 py-2">
            {t("страница_из", {
              current: page,
              total: Math.ceil(Number(totalResults) / 10),
            })}
          </span>
          <button
            className="bg-button hover:bg-button-hover text-white px-4 py-2 rounded transition-colors disabled:opacity-50"
            disabled={
              page >= Math.ceil(Number(totalResults) / 10) || isFetching
            }
            onClick={() => dispatch(paginationActions.setPage(page + 1))}
          >
            {t("вперед")}
          </button>
        </div>
      )}
    </>
  );
};

export default Pagination;
