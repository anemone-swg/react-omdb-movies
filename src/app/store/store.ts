import { combineReducers, configureStore } from "@reduxjs/toolkit";
import {
  FLUSH,
  PAUSE,
  PERSIST,
  persistReducer,
  persistStore,
  PURGE,
  REGISTER,
  REHYDRATE,
} from "redux-persist";
import storage from "redux-persist/lib/storage";
import { baseApi } from "@/shared/api/rtkApi";
import { paginationReducer } from "@/features/Pagination";

const rootReducer = combineReducers({
  pagination: paginationReducer,
  [baseApi.reducerPath]: baseApi.reducer,
});

const persistConfig = {
  key: "root",
  storage,
  blacklist: [baseApi.reducerPath],
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }).concat(baseApi.middleware),
});

export const persistor = persistStore(store);

export type AppState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
