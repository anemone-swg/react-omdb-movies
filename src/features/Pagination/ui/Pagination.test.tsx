import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { createReduxStore } from "@/app/store/store";
import Pagination from "./Pagination";

// в setupJest есть мок "react-i18next" для всех тестов,
// но тут мок конкретно для пагинации со своим ключом
jest.mock("react-i18next", () => ({
  useTranslation: () => ({
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    t: (key: string, options?: any) => {
      if (key === "страница_из")
        return `Страница ${options.current} из ${options.total}`;
      return key;
    },
  }),
}));

describe("Pagination", () => {
  test("render correctly with page = 5", () => {
    render(
      <Provider
        store={createReduxStore({
          pagination: { page: 5 },
        })}
      >
        <Pagination totalResults={"100"} />
      </Provider>,
    );
    const paginationDiv = screen.getByTestId("pagination-div");
    expect(paginationDiv).toBeInTheDocument();
    const paginationPage = screen.getByTestId("pagination-page");
    // expect(paginationPage).toHaveTextContent(/5\s*из/i);
    expect(paginationPage).toMatchSnapshot();
    screen.debug();
  });

  test("do not render", () => {
    render(
      <Provider store={createReduxStore()}>
        <Pagination totalResults={"100"} isFetching={true} />
      </Provider>,
    );
    expect(screen.queryByTestId("pagination-div")).not.toBeInTheDocument();
    screen.debug();
  });
});
