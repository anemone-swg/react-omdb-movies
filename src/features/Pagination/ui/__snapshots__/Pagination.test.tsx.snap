// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Pagination render correctly with page = 5 1`] = `
<span
  class="px-4 py-2"
  data-testid="pagination-page"
>
  Страница 5 из 10
</span>
`;

exports[`Pagination renders correctly with page = 5 1`] = `
<span
  class="px-4 py-2"
  data-testid="pagination-page"
>
  Страница 5 из 10
</span>
`;
